compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1188,6 +1192,74 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
         }
     }
 
+  /** Returns a block assigning to `targets` the corresponding values in `rhs`.
+   *
+   *  @param targets A sequence of untyped trees on the LHS of a multiple assignment.
+   *  @param rhs The RHS of a multuple assignment.
+   */
+  private def typedPairwiseAssignments(
+      targets: List[untpd.Tree], rhs: Tree
+  )(using Context): Tree =
+    val s = mutable.ListBuffer[untpd.Tree]()
+    val u = formPairwiseAssignments(
+      s, targets, untpd.TypedSplice(rhs), rhs.tpe,
+      EmptyTermName)
+    u.map((b) => typed(b)).getOrElse(unitLiteral)
+
+  /** Appends to `statements` the assignments of `targets` to corresponding values in `rhs`.
+   *
+   *  @param statements A partially constructed sequence representing the desugaring of a
+   *  multiple assignment.
+   *  @param targets A sequence of expression representing variables to assign.
+   *  @param rhs The expression of the values to assign..
+   *  @param rhsType The type of `rhs`.
+   *  @param prefix A prefix for the name of syntactic vals created by the methods.
+   */
+  private def formPairwiseAssignments(
+      statements: mutable.ListBuffer[untpd.Tree],
+      targets: List[untpd.Tree],
+      rhs: untpd.Tree,
+      rhsType: Type,
+      prefix: TermName
+  )(using Context): Option[untpd.Block] =
+    val source = UniqueName.fresh(prefix)
+    val d = untpd.ValDef(source, untpd.TypeTree(rhsType), rhs)
+      .withSpan(rhs.span)
+      .withFlags(Synthetic)
```

Comment:
The _first_ `rhs` is indeed in a `TypedSplice`. See the call in `typedPairwiseAssignments`. The problem is that subsequent right hand sides (which are passed in recursive calls) are not typed. They are the result of the local def `rhsElement`, which produces an untyped `Select`.

I couldn't find a way to get a typed version of this select during the recursion and figured that if it has to be typed anyway, then it would be acceptable to just have the typer re-visit the untyped tree once all transformation are complete. Trees encoding multiple assignment should be relatively shallow in practice.